src/layer/AtmosphereLayer.js

```diff
@@ -0,0 +1,414 @@
+                skyData.indicesVboCacheKey = dc.gpuResourceCache.generateCacheKey();
+            }
+
+            vboId = dc.gpuResourceCache.resourceForKey(skyData.indicesVboCacheKey);
+            if (!vboId) {
+                vboId = gl.createBuffer();
+                dc.gpuResourceCache.putResource(skyData.indicesVboCacheKey, vboId, this._skyTriStrip.length * 2);
+                skyData.refreshIndicesBuffer = true;
+            }
+
+            gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, vboId);
+            if (skyData.refreshIndicesBuffer) {
+                gl.bufferData(gl.ELEMENT_ARRAY_BUFFER,
+                    new Uint16Array(this._skyTriStrip), gl.STATIC_DRAW);
+                dc.frameStatistics.incrementVboLoadCount(1);
+                skyData.refreshIndicesBuffer = false;
+            }
+        };
+
+        // Internal. Intentionally not documented.
+        AtmosphereLayer.prototype.drawSky = function (dc) {
+
+            var gl = dc.currentGlContext,
+                program = dc.findAndBindProgram(SkyProgram);
+
+            program.loadGlobeRadius(gl, dc.globe.equatorialRadius);
+
+            program.loadEyePoint(gl, dc.navigatorState.eyePoint);
+
+            program.loadVertexOrigin(gl, Vec3.ZERO);
+
+            program.loadModelviewProjection(gl, dc.navigatorState.modelviewProjection);
+
+            program.loadLightDirection(gl, this.activeLightDirection);
+
+            program.setScale(gl);
+
+            gl.depthMask(false);
+            gl.frontFace(gl.CW);
+
+            this.setSkyPoints(dc, program.getAltitude());
+            this.setSkyTrianglesIndices();
+
+            this.applySkyVertices(dc);
+            this.applySkyIndices(dc);
+
+            gl.drawElements(gl.TRIANGLE_STRIP, this._skyTriStrip.length, gl.UNSIGNED_SHORT, 0);
+
+            gl.depthMask(true);
+            gl.frontFace(gl.CCW);
+        };
+
+        // Internal. Intentionally not documented.
+        AtmosphereLayer.prototype.drawGround = function (dc) {
+
+            var gl = dc.currentGlContext,
+                program = dc.findAndBindProgram(GroundProgram),
+                terrain = dc.terrain,
+                textureBound;
+
+            program.loadGlobeRadius(gl, dc.globe.equatorialRadius);
+
+            program.loadEyePoint(gl, dc.navigatorState.eyePoint);
+
+            program.loadLightDirection(gl, this.activeLightDirection);
+
+            program.setScale(gl);
+
+            // Use this layer's night image when the light location is different than the eye location.
+            if (this.nightImageSource && this.lightLocation) {
+                this._activeTexture = dc.gpuResourceCache.resourceForKey(this.nightImageSource);
+                if (!this._activeTexture) {
+                    this._activeTexture = dc.gpuResourceCache.retrieveTexture(gl, this.nightImageSource);
+                }
+                textureBound = this._activeTexture && this._activeTexture.bind(dc);
+            }
+
```

Comment:
It's not obvious, but this code accesses an internal property of the Tessellator, currentTiles, which could be empty if the tessellator were modified to avoid keeping tile references post-tessellation. Instead of terrain.tessellator.currentTiles, use terrain.surfaceGeometry, which is also a list of TerrainTile instances.